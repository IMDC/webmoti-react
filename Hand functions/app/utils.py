import logging
import signal
import sys
from logging.handlers import RotatingFileHandler
from time import sleep

from constants import LOG_PATH, SERVO_PIN

# from RPi import GPIO


def setup_gpio():
    # GPIO.setmode(GPIO.BOARD)
    # GPIO.setup(SERVO_PIN, GPIO.OUT)
    # pwm = GPIO.PWM(SERVO_PIN, 50)
    # pwm.start(0)

    pass


def setup_logging():
    log_format = "%(asctime)s - %(levelname)s - %(message)s"
    logging.basicConfig(level=logging.INFO, format=log_format, handlers=[])
    # rotate when log file is 5mb
    handler = RotatingFileHandler(LOG_PATH, maxBytes=5 * 1024 * 1024, backupCount=5)
    formatter = logging.Formatter(log_format)
    handler.setFormatter(formatter)
    logging.getLogger().addHandler(handler)


def setup_handlers():
    # setup cleanup handlers
    def signal_handler(*_):
        # pwm.stop()
        # GPIO.cleanup()
        logging.info("Cleaning up and exiting...")
        sys.exit(0)

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)


class ServoController:
    def __init__(self):
        # GPIO.setmode(GPIO.BOARD)
        # GPIO.setup(SERVO_PIN, GPIO.OUT)
        # self.pwm = GPIO.PWM(SERVO_PIN, 50)
        # self.pwm.start(0)
        pass

    def set_angle(self, angle):
        # Convert angle to duty cycle (2 to 12)
        # duty_cycle = (angle / 18) + 2
        # self.pwm.ChangeDutyCycle(duty_cycle)
        # sleep(1.5)

        print(angle)


servo_controller = ServoController()
