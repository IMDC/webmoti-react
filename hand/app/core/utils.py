import logging
import signal
import sys

from core.servo_controller import servo_controller


def setup_handlers():
    # setup cleanup handlers
    def signal_handler(*_):
        servo_controller.stop()
        logging.info("Cleaning up and exiting...")
        sys.exit(0)

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
