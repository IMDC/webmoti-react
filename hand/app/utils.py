import logging
import pathlib
import signal
import sys
from logging.handlers import RotatingFileHandler
from time import sleep

from RPi import GPIO

from constants import LOG_PATH, SERVO_PIN


def setup_logging():
    path = pathlib.Path(LOG_PATH)
    if not path.parent.is_dir():
        path = pathlib.Path(__file__).parent / "hand_server.log"

    log_format = "%(asctime)s - %(levelname)s - %(message)s"
    logging.basicConfig(level=logging.INFO, format=log_format, handlers=[])
    # rotate when log file is 5mb
    handler = RotatingFileHandler(path, maxBytes=5 * 1024 * 1024, backupCount=5)
    formatter = logging.Formatter(log_format)
    handler.setFormatter(formatter)
    logging.getLogger().addHandler(handler)


def setup_handlers():
    # setup cleanup handlers
    def signal_handler(*_):
        servo_controller.stop()
        logging.info("Cleaning up and exiting...")
        sys.exit(0)

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)


class ServoController:
    def __init__(self):
        GPIO.setmode(GPIO.BOARD)
        GPIO.setup(SERVO_PIN, GPIO.OUT)
        self.pwm = GPIO.PWM(SERVO_PIN, 50)
        self.pwm.start(0)

    def set_angle(self, angle):
        # Convert angle to duty cycle (2 to 12)
        duty_cycle = (angle / 18) + 2
        self.pwm.ChangeDutyCycle(duty_cycle)
        sleep(1.5)

    def stop(self):
        self.pwm.stop()
        GPIO.cleanup()


servo_controller = ServoController()
