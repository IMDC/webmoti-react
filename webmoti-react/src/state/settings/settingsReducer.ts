import { Track, VideoBandwidthProfileOptions } from 'twilio-video';

export interface Settings {
  trackSwitchOffMode: VideoBandwidthProfileOptions['trackSwitchOffMode'];
  dominantSpeakerPriority?: Track.Priority;
  bandwidthProfileMode: VideoBandwidthProfileOptions['mode'];
  maxAudioBitrate: string;
  contentPreferencesMode?: 'auto' | 'manual';
  clientTrackSwitchOffControl?: 'auto' | 'manual';
}

type SettingsKeys = keyof Settings;

export interface SettingsAction {
  name: SettingsKeys;
  value: string;
}

export const initialSettings: Settings = {
  // https://www.twilio.com/docs/video/tutorials/developing-high-quality-video-applications
  // https://www.twilio.com/docs/video/tutorials/using-bandwidth-profile-api#specifying-a-bw-profile

  // disabled: never turn off tracks when low bandwidth
  trackSwitchOffMode: 'disabled',
  // high: dominant speaker tracks will be high priority
  dominantSpeakerPriority: 'high',
  // presentation since there is a main speaker
  bandwidthProfileMode: 'presentation',
  maxAudioBitrate: '48000',
  contentPreferencesMode: 'auto',
  // manual: cameras don't turn off when zooming in to level 3 on board-view
  // (auto makes it turn off when zoomed because most of the zoomed video is not visible/rendered)
  clientTrackSwitchOffControl: 'manual',
};

// This inputLabels object is used by ConnectionOptions.tsx. It is used to populate the id, name, and label props
// of the various input elements. Using a typed object like this (instead of strings) eliminates the possibility
// of there being a typo.
export const inputLabels: Record<SettingsKeys, string> = Object.keys(initialSettings).reduce(
  (acc, key) => {
    acc[key as SettingsKeys] = key;
    return acc;
  },
  {} as Record<SettingsKeys, string>
);

export function settingsReducer(state: Settings, action: SettingsAction) {
  return {
    ...state,
    [action.name]: action.value === 'default' ? undefined : action.value,
  };
}
