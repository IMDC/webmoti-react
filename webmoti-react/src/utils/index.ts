import { Conversation, JSONObject, Message } from '@twilio/conversations';

import { WEBMOTI_CAMERA_1, WEBMOTI_CAMERA_2 } from '../constants';
import TTSMessage from '../components/ChatWindow/TTSMessage';

export const isMobile = (() => {
  if (typeof navigator === 'undefined' || typeof navigator.userAgent !== 'string') {
    return false;
  }
  return /Mobile/.test(navigator.userAgent);
})();

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && value.constructor === Object;
}

// Recursively removes any object keys with a value of undefined
export function removeUndefineds<T>(obj: T): T {
  if (!isPlainObject(obj)) return obj;

  const target: { [name: string]: unknown } = {};

  for (const key in obj) {
    const val = obj[key];
    if (typeof val !== 'undefined') {
      target[key] = removeUndefineds(val);
    }
  }

  return target as T;
}

export async function getDeviceInfo() {
  const devices = await navigator.mediaDevices.enumerateDevices();

  return {
    audioInputDevices: devices.filter((device) => device.kind === 'audioinput'),
    videoInputDevices: devices.filter((device) => device.kind === 'videoinput'),
    audioOutputDevices: devices.filter((device) => device.kind === 'audiooutput'),
    hasAudioInputDevices: devices.some((device) => device.kind === 'audioinput'),
    hasVideoInputDevices: devices.some((device) => device.kind === 'videoinput'),
  };
}

// This function will return 'true' when the specified permission has been denied by the user.
// If the API doesn't exist, or the query function returns an error, 'false' will be returned.
export async function isPermissionDenied(name: 'camera' | 'microphone') {
  const permissionName = name as PermissionName; // workaround for https://github.com/microsoft/TypeScript/issues/33923

  if (navigator.permissions) {
    try {
      const result = await navigator.permissions.query({ name: permissionName });
      return result.state === 'denied';
    } catch {
      return false;
    }
  } else {
    return false;
  }
}

export const isWebmotiVideo = (identity: string) => {
  return identity === WEBMOTI_CAMERA_1 || identity === WEBMOTI_CAMERA_2;
};

const getCustomMsgAttribute = (attribute: string) => JSON.stringify({ [attribute]: true });

export const sendSystemMsg = (conversation: Conversation | null, msg: string) => {
  // send with an attribute to differentiate from normal msg
  conversation?.sendMessage(msg, { attributes: getCustomMsgAttribute('systemMsg') });
};

export const sendSystemAudioMsg = (conversation: Conversation | null, message: TTSMessage) => {
  if (message.audioBlob === null) {
    return;
  }

  // send text message first
  conversation?.sendMessage(`TTS: ${message.text}`);

  // then send audio file
  const formData = new FormData();
  formData.append('media', message.audioBlob, 'TTS.mp3');
  formData.append('contentType', message.audioBlob.type);
  conversation?.sendMessage(formData, { attributes: getCustomMsgAttribute('ttsAudioMessage') });
};

const checkMessageAttribute = (message: Message, attribute: string): boolean => {
  const attrObj = message.attributes as JSONObject;

  if (attrObj.attributes === undefined) {
    // attribute isn't there
    return false;
  }

  try {
    const parsedAttributes = JSON.parse(attrObj.attributes as string);
    return parsedAttributes[attribute] === true;
  } catch {
    return false;
  }
};

export const checkSystemMsg = (message: Message) => {
  return checkMessageAttribute(message, 'systemMsg');
};

export const checkTTSAudioMsg = (message: Message) => {
  return checkMessageAttribute(message, 'ttsAudioMessage');
};
