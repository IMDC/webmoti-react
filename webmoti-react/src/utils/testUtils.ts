import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

export function renderWithUser(ui: React.ReactElement) {
  const user = userEvent.setup();
  return {
    user,
    ...render(ui),
  };
}
