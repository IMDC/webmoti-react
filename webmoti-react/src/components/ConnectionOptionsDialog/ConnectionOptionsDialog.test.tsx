import { render, screen, waitFor } from '@testing-library/react';
import { renderWithUser } from '../../utils/testUtils';

import ConnectionOptionsDialog from './ConnectionOptionsDialog';
import { initialSettings } from '../../state/settings/settingsReducer';
import { useAppState } from '../../state';
import useRoomState from '../../hooks/useRoomState/useRoomState';

jest.mock('../../hooks/useRoomState/useRoomState');
jest.mock('../../state');

const mockUseAppState = useAppState as jest.Mock<any>;
const mockUseRoomState = useRoomState as jest.Mock<any>;

const mockDispatchSetting = jest.fn();
mockUseAppState.mockImplementation(() => ({ settings: initialSettings, dispatchSetting: mockDispatchSetting }));

describe('the ConnectionOptionsDialog component', () => {
  afterEach(jest.clearAllMocks);

  describe('when not connected to a room', () => {
    beforeEach(() => {
      mockUseRoomState.mockImplementation(() => 'disconnected');
    });

    it('should render correctly', () => {
      const { container } = render(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
      expect(container).toMatchSnapshot();
    });

    it('should dispatch settings changes', async () => {
      const { user } = renderWithUser(<ConnectionOptionsDialog open={true} onClose={() => {}} />);

      const selectBox = screen.getByRole('combobox', { name: /dominant speaker priority/i });
      await user.click(selectBox);

      // select a different value than the default
      const newOption = screen.getByRole('option', { name: 'Low' });
      await user.click(newOption);

      await waitFor(() => {
        expect(mockDispatchSetting).toHaveBeenCalledWith({
          name: 'dominantSpeakerPriority',
          value: 'low',
        });
      });
    });

    it('should not dispatch settings changes from a number field when there are non-digits in the value', async () => {
      const { user } = renderWithUser(<ConnectionOptionsDialog open={true} onClose={() => {}} />);

      const input = screen.getByLabelText(/max audio bitrate/i);
      await user.clear(input);
      await user.type(input, '123456a');

      const calls = mockDispatchSetting.mock.calls;
      expect(calls).not.toContainEqual([{ name: 'maxAudioBitrate', value: '123456a' }]);
    });

    it('should dispatch settings changes from a number field when there are only digits in the value', async () => {
      const { user } = renderWithUser(<ConnectionOptionsDialog open={true} onClose={() => {}} />);

      const input = screen.getByLabelText(/max audio bitrate/i);
      await user.clear(input);
      await user.type(input, '123456');

      expect(mockDispatchSetting).toHaveBeenCalledWith({
        name: 'maxAudioBitrate',
        value: '480006',
      });
    });

    describe('when connected to a room', () => {
      mockUseRoomState.mockImplementation(() => 'connected');
      it('should render correctly', () => {
        const { container } = render(<ConnectionOptionsDialog open={true} onClose={() => {}} />);
        expect(container).toMatchSnapshot();
      });
    });
  });
});
