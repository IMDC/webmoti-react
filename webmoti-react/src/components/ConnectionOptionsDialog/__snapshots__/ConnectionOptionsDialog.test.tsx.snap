// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`the ConnectionOptionsDialog component > when not connected to a room > should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`the ConnectionOptionsDialog component > when not connected to a room > when connected to a room > should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`the ConnectionOptionsDialog component when not connected to a room should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;

exports[`the ConnectionOptionsDialog component when not connected to a room when connected to a room should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
